File: python/2_sum.py
from typing import List 

def twoSum(nums: List[int], target: int) -> List[int]:
    memo = {}
    for i, n in enumerate(nums):
        if(n in memo):
            return [memo[n], i]
        
        complement = target - n
        memo[complement] = i
    return []


assert twoSum([1,2,3], 4) == [0,2], "example"
assert twoSum([1,2,3], 10) == [], "negative example"
assert twoSum([1,2,3,-1], 1) == [1,3], "negative numbers"

File: csharp/2_sum.cs
// Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

// You may assume that each input would have exactly one solution, and you may not use the same element twice.

// You can return the answer in any order.

class Solution {
    // space complexity O(n)
    // memory O(n)
    public int[] TwoSum(int[] nums, int target) {
        var remainderLookup = new Dictionary<int, int>();
        for(int i = 0; i<nums.Length; ++i) {
            var n = nums[i];
            var targetN = target - n;
            if(remainderLookup.ContainsKey(targetN)) return new int[]{remainderLookup[targetN], i};
            else{
                remainderLookup[n] = i;
            }
        }
        return null;
    }
}